File: training-data/ruby/merge_intervals.rb
def merge_intervals(intervals)
  sorted = intervals.sort_by(&:first)
  merged = [sorted.first]

  sorted.drop(1).each do |current|
    last = merged.last
    if current.first <= last.last
      last[1] = [last.last, current.last].max
    else
      merged << current
    end
  end

  merged
end

puts merge_intervals([[1, 3], [2, 6], [8, 10], [15, 18]]).inspect
puts merge_intervals([[1, 4], [4, 5]]).inspect

File: training-data/rust/merge_intervals.rs
#[derive(Debug, Clone, Copy)]
struct Interval {
    start: i32,
    end: i32,
}

fn merge_intervals(mut intervals: Vec<Interval>) -> Vec<Interval> {
    intervals.sort_by_key(|iv| iv.start);

    intervals.into_iter().fold(Vec::new(), |mut merged, current| {
        match merged.last_mut() {
            Some(last) if current.start <= last.end => {
                last.end = last.end.max(current.end);
            }
            _ => merged.push(current),
        }
        merged
    })
}

fn main() {
    let intervals = vec![
        Interval { start: 1, end: 3 },
        Interval { start: 2, end: 6 },
        Interval { start: 8, end: 10 },
        Interval { start: 15, end: 18 },
    ];
    println!("{:?}", merge_intervals(intervals));
}
